File: src/vodoo/aio/knowledge.py
"""Async knowledge article operations for Vodoo."""

from typing import Any

from vodoo.aio._domain import AsyncDomainNamespace
from vodoo.knowledge import _build_article_values, _KnowledgeAttrs


class AsyncKnowledgeNamespace(_KnowledgeAttrs, AsyncDomainNamespace):
    """Async namespace for knowledge.article model."""

    async def create(
        self,
        name: str,
        *,
        body: str | None = None,
        parent_id: int | None = None,
        category: str | None = None,
        icon: str | None = None,
        **extra_fields: Any,
    ) -> int:
        """Create a knowledge article."""
        values = _build_article_values(
            name,
            body=body,
            parent_id=parent_id,
            category=category,
            icon=icon,
            **extra_fields,
        )
        return await self._client.create(self._model, values)

    async def url(self, record_id: int) -> str:  # type: ignore[override]
        """Get the web URL for a knowledge article.

        Tries the ``article_url`` field first, falls back to the standard URL.
        """
        article = await self.get(record_id, fields=["article_url"])
        if article.get("article_url"):
            return str(article["article_url"])
        return super().url(record_id)

File: packages/typescript/src/namespaces/knowledge.ts
import type { OdooClientApi } from "../client-api.js";
import { richTextToHtml } from "../content.js";
import type { RichText } from "../content.js";
import { GeneratedKnowledgeNamespace } from "../generated/knowledge.js";
import type { OdooRecord } from "../types.js";
import type { ListOptions } from "./domain.js";

export interface KnowledgeRecord extends OdooRecord {
  id: number;
  name: string;
  write_date?: Date | null;
}

export interface CreateArticleOptions {
  body?: RichText;
  parentId?: number;
  category?: string;
  icon?: string;
  extraFields?: Readonly<Record<string, unknown>>;
}

export class KnowledgeNamespace extends GeneratedKnowledgeNamespace {
  constructor(client: OdooClientApi) {
    super(client);
  }

  override async list(options: ListOptions = {}): Promise<KnowledgeRecord[]> {
    return (await super.list(options)) as KnowledgeRecord[];
  }

  override async get(
    recordId: number,
    fields?: readonly string[],
  ): Promise<KnowledgeRecord> {
    return (await super.get(recordId, fields)) as KnowledgeRecord;
  }

  create(name: string, options: CreateArticleOptions = {}): Promise<number> {
    const values: Record<string, unknown> = {
      name,
      ...options.extraFields,
    };
    if (options.body !== undefined) {
      values.body = richTextToHtml(options.body);
    }
    if (options.parentId !== undefined) values.parent_id = options.parentId;
    if (options.category !== undefined) values.category = options.category;
    if (options.icon !== undefined) values.icon = options.icon;
    return this.client.create("knowledge.article", values);
  }

  /** Async TypeScript adaptation of Python's article_url lookup. */
  async resolveUrl(recordId: number): Promise<string> {
    const article = await this.get(recordId, ["article_url"]);
    return typeof article.article_url === "string" && article.article_url !== ""
      ? article.article_url
      : this.url(recordId);
  }
}
